sympy/core/operations.py

```diff
@@ -49,15 +53,74 @@ def __new__(cls, *args, **options):
         if len(args) == 1:
             return args[0]
 
-        c_part, nc_part, order_symbols = cls.flatten(args)
+        c_part, nc_part = cls.flatten(args)
         is_commutative = not nc_part
         obj = cls._from_args(c_part + nc_part, is_commutative)
         obj = cls._exec_constructor_postprocessors(obj)
-
-        if order_symbols is not None:
-            return Order(obj, *order_symbols)
         return obj
 
+    @classmethod
+    def _constructor_hook(cls, *args, ignore_hook=False, **options):
+        # Select the hook from args, apply args and options to it and return the result.
+        # cls.dispatcher must be defined which returns the hook function of given argument and
+        # None for unregistered type.
+        # Also, If None is returned from this method, hook is not applied.
+        # See https://github.com/sympy/sympy/pull/18769
+
+        # ignore the hook
+        if ignore_hook:
+            return None
+
+        _, _, hook = cls._find_hook(cls, *args)
+        if hook is None:     # No hook is defined for args, so hook cannot be applied.
+            return None
+        else:
+            return hook(*args, **options)
+
+    @classmethod
+    def _find_hook(cls, *args):
+        # Helper for _constructor_hook. This method selects the highst hook and the class which has it.
+        # Returns: (func, op_priority of func, hook of func)
+
+        dispatcher = cls.__dict__.get('dispatcher') # do not allow inheritance of dispatcher
+        if dispatcher is None:  # No hook is applied if cls isn't designed for hook
+            return None, None, None
+
+        hooks = []
+        for a in args:
+            if isinstance(a, cls):  # To find hook in e.g. Add(1, Order(x))
+                func, op_priority, hook = cls._find_hook(*a.args)
```

Comment:
To be clear, this code path is hit for the inner `Add` in `Add(Add(1, Order(x), 1)`?